### PR TITLE
feat: implement Puffin format

### PR DESCRIPTION
**Pull request by zhongzc**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Part of #2705

implement {sync|async} {reader|writer} for Puffin.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/puffin/src/tests/mod.rs
Comment (waynexia) [inline src/puffin/src/tests/mod.rs:1]: Why this is binary

Comment (zhongzc) [reply src/puffin/src/tests/mod.rs:1]: github knows

Comment (zhongzc) [reply src/puffin/src/tests/mod.rs:1]: .dev works fine

Comment (evenyag) [reply src/puffin/src/tests/mod.rs:1]: Why does the `mod.rs` occupy 8.25KB?

By the way, can we remove `mod.rs`? Most crates in the project don't use `mod.rs`:
https://doc.rust-lang.org/edition-guide/rust-2018/path-changes.html#no-more-modrs

**Review by evenyag — COMMENTED** [review]
Should we create a new repo for the puffin crate? We can create a repo under our team. @waynexia @v0y4g3r 

This PR is too large and I'd suggest breaking it into 3 or 4 small PRs that each is less than 1k line
- init crate and file formats
- writer
- reader
- partial reader

**Comment by zhongzc** [comment]
> Why does the `mod.rs` occupy 8.25KB?

It's average. 40B/line * 200lines = 8KB


> can we remove `mod.rs`?

Sure.


> Should we create a new repo for the puffin crate

We can move out when the implementation becomes more mature and stable. Multiple repos will affect development efficiency.


> breaking it into 3 or 4 small PRs

good suggestion

**Comment by v0y4g3r** [comment]
Maybe we should create a `puffin-rs` crate?

**Comment by zhongzc** [comment]
> Maybe we should create a `puffin-rs` crate?

Do you mean creating a new repo?

**Pull request closed without merging**